mathext/hyp2f1.go

```diff
@@ -0,0 +1,466 @@
+package mathext
+
+import (
+	"fmt"
+	"math"
+)
+
+// Machine epsilon used in Cephes.
+var mACHEP float64 = (math.Nextafter(1.0, 2.0) - 1.0) / 2
+
+const eps float64 = 1e-13
+const ethresh float64 = 1e-12
+
+// Hyp2f1 returns the value of the Gaussian Hypergeometric function at z.
+// For |z| < 1, this implementation follows the Cephes library.
+// For |z| > 1, this implementation perform analytic continuation via relevant Hypergeometric identities.
+// See https://en.wikipedia.org/wiki/Hypergeometric_function for more details.
+func Hyp2f1(a float64, b float64, c float64, z float64) (float64, error) {
+	if math.Abs(z) < 1 {
+		return hyp2f1(a, b, c, z)
+	}
+
+	// Function undefined between the 1 and Inf branch points.
+	if z > 0 {
+		return math.NaN(), fmt.Errorf("analytic continuation failed")
+	}
+
+	// Analytic continuation formula contains infinities from Gamma(a-b) when a == b.
+	// Make a and b different using equation 15.3.4 from Abramowitz.
+	// M. Abramowitz and I. A. Stegun 1965. Handbook of Mathematical Functions, New York: Dover.
+	if a == b {
+		y, err := Hyp2f1(a, c-b, c, z/(z-1))
+		if err != nil {
+			return y, err
+		}
+		return math.Pow(1-z, -a) * y, nil
+	}
+
+	// Analytic continuation based on https://www.johndcook.com/blog/2021/11/03/escaping-the-unit-disk/
+	y1, err := hyp2f1(a, 1-c+a, 1-b+a, 1/z)
+	if err != nil {
+		return y1, err
+	}
+	y1 *= math.Gamma(c) / math.Gamma(b) * math.Gamma(b-a) / math.Gamma(c-a) * math.Pow(-z, -a)
+	y2, err := hyp2f1(b, 1-c+b, 1-a+b, 1/z)
+	if err != nil {
+		return y2, err
+	}
+	y2 *= math.Gamma(c) / math.Gamma(a) * math.Gamma(a-b) / math.Gamma(c-b) * math.Pow(-z, -b)
+	return y1 + y2, nil
+}
+
+/*							hyp2f1.c
```

Comment:
The section below this should live in mathext/internal/cephes where it will be clearly marked as licensed under the Cephes license. Best in its own file with appropriate license specialisation.

This will also allow sharing of precision constants.

It would be nice if fewer fortranish calling conventions were used, but this can be fixed later.

Reply:
Yes, this should really go into internal/cephes, sorry for not being aware of the folder structures.
Putting this into internal/cephes really brings lots of benefits, such as reusing of constants and error messages like you mentioned.